text/0000-default-field-values.md

```diff
@@ -0,0 +1,1979 @@
+```
+
+Then, `RecordField` is changed into:
+
+```rust
+RecordField = attrs:OuterAttr* vis:Vis? name:IDENT ":" ty:Type { "=" def:Expr }?;
+```
+
+Further, given the following partial definition for the expression grammar:
+
+```rust
+Expr = attrs:OuterAttr* kind:ExprKind;
+ExprKind =
+  | ...
+  | Struct:{ path:Path "{" attrs:InnerAttr* fields:StructExprFieldsAndBase "}" }
+  ;
+
+StructExprFieldsAndBase =
+  | Fields:{ fields:StructExprField* % "," ","? }
+  | Base:{ ".." base:Expr }
+  | FieldsAndBase:{ fields:StructExprField+ % "," "," ".." base:Expr }
+  ;
+StructExprField = attrs:OuterAttr* kind:StructExprFieldKind;
+StructExprFieldKind =
+  | Shorthand:IDENT
+  | Explicit:{ field:FieldName ":" expr:Expr }
+  ;
+```
+
+the rule `StructExprFieldsAndBase` is extended with:
+
+```rust
+StructExprFieldsAndBase =| FieldsAndDefault:{ fields:StructExprField+ % "," "," ".." };
+StructExprFieldsAndBase =| Default:{ ".." }
+```
+
+### Static semantics
+
+#### Defining defaults
+
+Given a `RecordField` where the default is specified, i.e.:
+
+```rust
+RecordField = attrs:OuterAttr* vis:Vis? name:IDENT ":" ty:Type "=" def:Expr;
+```
+
+all the following rules apply when type-checking:
+
+1. The expression `def` must be a constant expression.
+
+2. The expression `def` must coerce to the type `ty`.
+
+3. Generic parameters of the current items are accessible
+   ```rust
+   struct Bar<const A: usize> {
+       field: usize = A,
+   }
+   ```
+
+4. Default const expressions are *not* evaluated at definition time, only
+   during instantiation. This means that the following will not fail to compile:
+   ```rust
+   struct Bar {
+       field1: usize = panic!(),
+       field2: usize = 42,
+   }
+
+   let _ = Bar { field1: 0, .. };
+   ```
+   Having said that, it can be possible to proactivelly attempt to evaluate the
+   default values and emit a lint in a case where the expression is assured to always
+   fail (which would only be possible for expressions that do not reference `const`
+   parameters).
+
+5. The `struct`'s parameters are properly propagated, meaning the following is
+   possible:
+   ```rust
+   struct Bar<T> {
+       field: Vec<T> = Vec::new(),
+   }
+
+   let _ = Bar::<i32> { .. };
+   ```
+
+When lints check attributes such as `#[allow(lint_name)]` are placed on a
+`RecordField`, it also applies to `def` if it exists.
+
+#### Initialization expressions
+
+`Path { fields, .. }` is `const` since the defaulted fields are initialized
+from constants.
+
+## `#[derive(Default)]`
+
+When generating an implementation of `Default` for a `struct` named `$s` on
+which `#[derive(Default)]` has been attached, the compiler will omit all fields
+which have default values provided in the `struct`. The the associated function
+`default` shall then be defined as (where `$f_i` denotes the `i`-th field of
+`$s`):
+
+```rust
+fn default() -> Self {
+    $s { $f_i: Default::default(), .. }
+}
+```
+
+# Drawbacks
```

Comment:
In this section it is a bit unclear where each downsides ends and the next starts. Please add more explicit structure, e.g. with an enumeration or by starting each now downside with a separate header.